-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ef79e/13912/summary.html
**COMMIT**: fffebf70bb20a659d576ca96b1d80851369a9fee
**CMSSW**: CMSSW_11_3_X_2021-03-31-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-externals/llvm-project/11/13912/install.sh` to create a dev area with all the needed externals and cmssw changes.


## External Build

I found compilation error when building: 

<pre>/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/llvm/11.1.0-4ba386fa44c6eecfa07b9de39cf23cf0/llvm-11.1.0-6afd8d28e7729e1256cc832c60e4be6830f05dc6/clang-tools-extra/clang-tidy/cms/HandleCheck.cpp:164:10: warning: unused variable 'implicitObjectExpr' [-Wunused-variable]
164 |     auto implicitObjectExpr = matchedCallExpr->getImplicitObjectArgument();
|          ^~~~~~~~~~~~~~~~~~
[3964/4743] /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc900/external/gcc/9.3.0/bin/c++  -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/clang/tools/extra/clang-tidy/readability -I/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/llvm/11.1.0-4ba386fa44c6eecfa07b9de39cf23cf0/llvm-11.1.0-6afd8d28e7729e1256cc832c60e4be6830f05dc6/clang-tools-extra/clang-tidy/readability -I/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/llvm/11.1.0-4ba386fa44c6eecfa07b9de39cf23cf0/llvm-11.1.0-6afd8d28e7729e1256cc832c60e4be6830f05dc6/clang/include -Itools/clang/include -Iinclude -I/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/llvm/11.1.0-4ba386fa44c6eecfa07b9de39cf23cf0/llvm-11.1.0-6afd8d28e7729e1256cc832c60e4be6830f05dc6/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -O3 -DNDEBUG     -std=c++14 -MD -MT tools/clang/tools/extra/clang-tidy/readability/CMakeFiles/obj.clangTidyReadabilityModule.dir/SimplifyBooleanExprCheck.cpp.o -MF tools/clang/tools/extra/clang-tidy/readability/CMakeFiles/obj.clangTidyReadabilityModule.dir/SimplifyBooleanExprCheck.cpp.o.d -o tools/clang/tools/extra/clang-tidy/readability/CMakeFiles/obj.clangTidyReadabilityModule.dir/SimplifyBooleanExprCheck.cpp.o -c /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/llvm/11.1.0-4ba386fa44c6eecfa07b9de39cf23cf0/llvm-11.1.0-6afd8d28e7729e1256cc832c60e4be6830f05dc6/clang-tools-extra/clang-tidy/readability/SimplifyBooleanExprCheck.cpp
ninja: build stopped: subcommand failed.
error: Bad exit status from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.57cniu (%build)


RPM build errors:
Macro %rpmbuild_libdir defined but not used within scope
Bad exit status from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.57cniu (%build)
</pre>

--- a/clang-tools-extra/clang-tidy/cms/ESRecordGetCheck.cpp
+++ b/clang-tools-extra/clang-tidy/cms/ESRecordGetCheck.cpp
@@ -17,31 +17,13 @@ namespace clang {
 namespace tidy {
 namespace cms {
 
-const std::string esgettoken = "ESGetToken";
-const std::string eshandle = "ESHandle";
-const std::string esrecord = "EventSetupRecord";
-const std::string get = "get";
-const std::string thisp = "this->";
-
 
 void ESRecordGetCheck::registerMatchers(MatchFinder *Finder) {
-//  auto edmESGetToken = cxxRecordDecl(hasName("edm::ESGetToken"));
-//  auto edmESHandle = cxxRecordDecl(hasName("edm::ESHandle"));
-//  auto edmEventSetup = cxxRecordDecl(hasName("edm::EventSetup"));
-//  auto edmESRecord = cxxRecordDecl(hasName("edm::EventSetupReord"));
-//
-//  auto edmESHandleVarRef = declRefExpr(
-//                           hasDeclaration(varDecl()),
-//                           hasType(edmESHandle));
-//  auto edmGetTokenRef = declRefExpr(
-//                           hasDeclaration(varDecl()),
-//                           hasType(edmESGetToken));
-//  auto edmEventSetupRef = declRefExpr(
-//                       hasDeclaration(varDecl()),
-//                       hasType(edmEventSetup));
-//  
+
+  auto edmESGetToken = cxxRecordDecl(hasName("edm::ESGetToken"));
+
   auto ESRecord = cxxRecordDecl(
-                            isSameOrDerivedFrom("EventSetupRecord")
+                            isSameOrDerivedFrom(hasName("EventSetupRecord"))
                           );
 
   auto ESRgetDecl = cxxMethodDecl(
@@ -50,7 +32,8 @@ void ESRecordGetCheck::registerMatchers(MatchFinder *Finder) {
                          );
 
   auto getCall = cxxMemberCallExpr(
-                          callee(ESRgetDecl)
+                          callee(ESRgetDecl),
+                          unless(hasArgument(0,expr(hasType(edmESGetToken))))
                         ).bind("getcallexpr");
 
   Finder->addMatcher(getCall,this);

--- a/clang-tools-extra/test/clang-tidy/checkers/cms-eventsetuprecord-get.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/cms-eventsetuprecord-get.cpp
@@ -36,6 +36,21 @@ namespace edm {
     bool get(std::string const&, ESHandle<T>&);
 
   };
+
+  template <typename T>
+  class EventSetupRecordImplementation : public EventSetupRecord {
+  public:
+
+    using EventSetupRecord::get;
+
+    template <typename PRODUCT>
+    PRODUCT const& get(ESGetToken<PRODUCT, T> const & iToken) const;
+
+
+    template <typename PRODUCT>
+    PRODUCT const& get(ESGetToken<PRODUCT, T>& iToken) const;
+
+  }; 
   
   class EventSetup {
   public:
@@ -51,6 +66,9 @@ namespace edm {
 
 struct FooR : public edm::EventSetupRecord {};
 struct FooP {};
+
+class ESFooR : public edm::EventSetupRecordImplementation<ESFooR> {};
+
 
 class Bar {
 public:
@@ -94,6 +112,19 @@ bool Bar::doWork(edm::EventSetup& iSetup, edm::ESGetToken<FooP, FooR> const& tok
 // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: direct call of function EventSetupRecord::get(ESHandle&) is deprecated and should be replaced with a call to EventSetup::getHandle(ESGetToken&). To use ESGetToken see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module To get data with the token see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit [cms-esrget]
   pFooR->get("test", handle);
 // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: direct call of function EventSetupRecord::get(ESHandle&) is deprecated and should be replaced with a call to EventSetup::getHandle(ESGetToken&). To use ESGetToken see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module To get data with the token see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit [cms-esrget]
+
+  return true;
+}
+
+class Baz {
+  public:
+    edm::ESGetToken<FooR, ESFooR> m_token;
+    bool doWork( ESFooR const& record);
+};
+
+bool Baz::doWork( ESFooR const& record) {
+
+  const auto& esFooR = record.get(m_token);
 
   return true;
 }

